common/context/src/main/java/io/helidon/common/context/ContextSingleton.java

```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright (c) 2024 Oracle and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.helidon.common.context;
+
+import java.util.NoSuchElementException;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.function.Supplier;
+
+/**
+ * A context based "singleton" value holder that provides an indirection mechanism for static values.
+ * "Singleton" values are stored and resolved from a context classified with {@value #STATIC_CONTEXT_CLASSIFIER}
+ * in the {@link io.helidon.common.context.Contexts#context()} current context},
+ * or the {@link io.helidon.common.context.Contexts#globalContext() global context if none is defined}.
+ *
+ * @param <T> type of the value stored in context
+ */
+public final class ContextSingleton<T> {
+    /**
+     * Classifier used to register a context that is to serve as the static context.
+     */
+    public static final String STATIC_CONTEXT_CLASSIFIER = "helidon-singleton-context";
```

Comment:
```java
ContextSingleton.CLASSIFIER
```

is better than

```java
ContextSingleton.STATIC_CONTEXT_CLASSIFIER
```

At the very least we should not use "static", so either `SINGLETON_CONTEXT_CLASSIFIER`, `CONTEXT_CLASSIFIER` or just `CLASSIFIER`.